File: Top_Google_Questions/346._Moving_Average_from_Data_Stream/solution.py
class MovingAverage:

    def __init__(self, size: int):
        """
        Initialize your data structure here.
        """
        self.moves = [0] * size
        self.curr_size = 0
        self.max_size = size
        self.idx = 0
        self.total = 0

    def next(self, val: int) -> float:
        self.total -= self.moves[self.idx]
        self.total += val
        self.moves[self.idx] = val
        self.idx = (self.idx + 1) % self.max_size
        
        if self.curr_size < self.max_size:
            self.curr_size += 1
        
        return self.total / self.curr_size
            


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)

File: Top_Google_Questions/346._Moving_Average_from_Data_Stream/solution.js
/**
 * Initialize your data structure here.
 * @param {number} size
 */
var MovingAverage = function(size) {
    this.moves = new Array(size).fill(0);
    this.curr_size = 0;
    this.max_size = size;
    this.total = 0;
    this.idx = 0;
};

/** 
 * @param {number} val
 * @return {number}
 */
MovingAverage.prototype.next = function(val) {
    this.total -= this.moves[this.idx];
    this.total += val;
    this.moves[this.idx] = val;
    this.idx = (this.idx + 1) % this.max_size;
    
    if (this.curr_size < this.max_size) {
        this.curr_size += 1;
    }
    
    return this.total / this.curr_size;
};

/** 
 * Your MovingAverage object will be instantiated and called as such:
 * var obj = new MovingAverage(size)
 * var param_1 = obj.next(val)
 */